### PR TITLE
New package: encrypted-dns

### PR DESCRIPTION
**Pull request by MeganerdNL**
**DNSCrypt v2 server with support for DNSSEC, DoH forwarding and anonymized DNSCrypt**

#### Testing the changes
- I tested the changes in this PR: **YES**

#### New package
- This new package conforms to the [package requirements](https://github.com/void-linux/void-packages/blob/master/CONTRIBUTING.md#package-requirements): **YES**

#### Local build testing
- I built this PR locally for my native architecture, (x86_64-glibc)
- I built this PR locally for these architectures (if supported. mark crossbuilds):
  - x86_64-musl
  - i686-glibc
  - aarch64-glibc
  - aarch64-musl
  - armv7l-glibc
  - armv7l-musl
  - armv6l-glibc
  - armv6l-musl

### CODE REVIEW

File: srcpkgs/encrypted-dns/files/encrypted-dns.toml
Comment (classabbyamp) [inline srcpkgs/encrypted-dns/files/encrypted-dns.toml:1]: don't vendor this file, grab it from the builddir in post_install

File: srcpkgs/encrypted-dns/template [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+case "$XBPS_TARGET_MACHINE" in
+	*-musl) makedepends+=" musl-devel libstdc++";;
+esac
+
+post_install() {
+	vlicense LICENSE
+
+	# Install config file and list of undelegated TLDs
+	vinstall ${FILESDIR}/encrypted-dns.toml 0644 etc/encrypted-dns encrypted-dns.toml
```
Comment (classabbyamp) [inline srcpkgs/encrypted-dns/template:34]:
```suggestion
	vsconf example-encrypted-dns.toml
```
this is an example config so it should probably not be installed in /etc, especially when it contains things that should be randomly generated by the user

File: srcpkgs/encrypted-dns/template [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+	vinstall undelegated.txt 0644 var/lib/encrypted-dns undelegated.txt
```
Comment (classabbyamp) [inline srcpkgs/encrypted-dns/template:35]:
```suggestion
	vinstall undelegated.txt 0644 var/lib/encrypted-dns
```
redundant

File: srcpkgs/encrypted-dns/template [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+mutable_files="/var/lib/encrypted-dns/undelegated.txt"
+
+if [ "$XBPS_TARGET_NO_ATOMIC8" ]; then
+	XBPS_CROSS_RUSTFLAGS+=" -latomic"
+	makedepends+=" libatomic-devel"
+fi
+
+case "$XBPS_TARGET_MACHINE" in
+	*-musl) makedepends+=" musl-devel libstdc++";;
```
Comment (classabbyamp) [inline srcpkgs/encrypted-dns/template:27]: why is this needed?

File: srcpkgs/encrypted-dns/INSTALL [hunk 4]
```diff
@@ -0,0 +1,10 @@
+case "$ACTION" in
+    post)
+        if [ "$UPDATE" = "yes" ] ; then
+            :
+        else
+            chown _encrypted_dns:_encrypted_dns var/lib/encrypted-dns
+            chmod 750 var/lib/encrypted-dns
```
Comment (classabbyamp) [inline srcpkgs/encrypted-dns/INSTALL:7]: use `make_dirs`, but this may already be the default for a system account's homedir

File: srcpkgs/encrypted-dns/template [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+homepage="https://github.com/DNSCrypt/encrypted-dns-server"
+distfiles="https://github.com/DNSCrypt/encrypted-dns-server/archive/refs/tags/${version}.tar.gz"
+checksum=5e4f9143313bf58888c31ec4e220e9fb65b28b60fe5b6aff872f9f2ecb7537d5
+
+system_accounts="_encrypted_dns"
+_encrypted_dns_homedir="/var/lib/encrypted-dns"
+
+conf_files="/etc/encrypted-dns/encrypted-dns.toml"
+
+mutable_files="/var/lib/encrypted-dns/undelegated.txt"
```
Comment (classabbyamp) [inline srcpkgs/encrypted-dns/template:19]:
```suggestion
system_accounts="_encrypted_dns"
_encrypted_dns_homedir="/var/lib/encrypted-dns"
conf_files="/etc/encrypted-dns/encrypted-dns.toml"
mutable_files="/var/lib/encrypted-dns/undelegated.txt"
```

File: srcpkgs/encrypted-dns/files/encrypted-dns.toml
Comment (MeganerdNL) [reply srcpkgs/encrypted-dns/files/encrypted-dns.toml:1]: > don't vendor this file, grab it from the builddir in post_install

I can do that, actually did it at first when creating this but then decided this (saw some other packages do it).

Is it ok to change (or add in) some values like _user/group_ in the example config via _vsed_  in the template to accommodate? I have the code here already.

File: srcpkgs/encrypted-dns/template (hunk 3 above)
Comment (MeganerdNL) [reply srcpkgs/encrypted-dns/template:27]: > why is this needed?

It won't compile on musl otherwise. 
Complains of missing dependencies that are not there on glibc.

File: srcpkgs/encrypted-dns/INSTALL (hunk 4 above)
Comment (MeganerdNL) [reply srcpkgs/encrypted-dns/INSTALL]: chown is indeed not needed, removed.
chmod is still needed, because /var/lib/encrypted-dns should not be world-readable per package documentation.
make_dirs does not work here, because _encrypted_dns_homedir still installes as 755 and make_dirs will not override this.

File: srcpkgs/encrypted-dns/template (hunk 5 above)
Comment (MeganerdNL) [reply srcpkgs/encrypted-dns/template]: done.

File: srcpkgs/encrypted-dns/template (hunk 1 above)
Comment (MeganerdNL) [reply srcpkgs/encrypted-dns/template]: done.

File: srcpkgs/encrypted-dns/template (hunk 2 above)
Comment (MeganerdNL) [reply srcpkgs/encrypted-dns/template]: done.

File: srcpkgs/encrypted-dns/template (hunk 3 above)
Comment (classabbyamp) [reply srcpkgs/encrypted-dns/template:23]: works fine with a musl masterdir:

```
$ ./xbps-src -m masterdir-musl -a armv6l-musl pkg encrypted-dns
...
=> Registering new packages to /host/binpkgs/encrypted-dns
index: added `encrypted-dns-0.9.13_1' (armv6l-musl).
index: 1 packages registered.
```

File: srcpkgs/encrypted-dns/template [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+case "$XBPS_TARGET_MACHINE" in
+	*-musl) makedepends+=" musl-devel libstdc++";;
+esac
+
+post_install() {
+	vlicense LICENSE
+	vinstall undelegated.txt 0644 var/lib/encrypted-dns
+	vsconf example-encrypted-dns.toml
+	vmkdir etc/encrypted-dns
+	touch ${DESTDIR}/etc/encrypted-dns/.placeholder
```
Comment (classabbyamp) [inline srcpkgs/encrypted-dns/template:31]:
```suggestion
```
the user can create the dir themselves (and `make_dirs=` is the proper way to do this anyways)

**Comment by MeganerdNL** [comment]
@classabbyamp

Ready for review again.

- Regarding these lines in _template_:

```
if [ "$XBPS_TARGET_NO_ATOMIC8" ]; then
	XBPS_CROSS_RUSTFLAGS+=" -latomic"
	makedepends+=" libatomic-devel"
fi

case "$XBPS_TARGET_MACHINE" in
	*-musl) makedepends+=" musl-devel libstdc++";;
esac
```

You said it build fine with:

> ./xbps-src -m masterdir-musl -a armv6l-musl pkg encrypted-dns

But I cannot reproduce that on my machine.
Also: I tried to remove both or one of the blocks _each_ and they always fail to compile in **Github on the musl-builds too**. So not only on my machine when cross-compiling.  There are several other packages that need these blocks, or one of them. Let me know how to fix it otherwise, because I cannot make it work without these blocks.

- As with the make_dirs remark: you were of course right. I had to change the order of _make_dirs_ and __homedir_  and then the mode (750) was correct after installing.

- Made a files/README.voidlinux file instead of INSTALL.msg, although I  think that the creation of a mandatory configfile that has an example is quite important. There are plenty of packages with far less important information in that file, but maybe that's something that slipped in in the past.

Thanks.

File: srcpkgs/encrypted-dns/template (hunk 3 above)
Comment (MeganerdNL) [reply srcpkgs/encrypted-dns/template:24]: Not reproducible.  Also failing on Github. See comment in main section.

**Comment by classabbyamp** [comment]
because of recent changes to the cross toolchains, only this is needed:
```diff
  if [ "$XBPS_TARGET_NO_ATOMIC8" ]; then
  	XBPS_CROSS_RUSTFLAGS+=" -latomic"
- 	makedepends+=" libatomic-devel"
  fi
- 
- case "$XBPS_TARGET_MACHINE" in
- 	*-musl) makedepends+=" musl-devel libstdc++";;
- esac
```
(tested with a musl masterdir (`xbps-src -m masterdir-musl binary-bootstrap x86_64-musl` and `xbps-src -m masterdir-musl -a armv6l-musl pkg encrypted-dns`))

> I think that the creation of a mandatory configfile that has an example is quite important

by important I mean things like breaking changes (see pipewire, for example). things like optional dependencies or getting-started steps are not important enough and just add more noise that users will probably ignore.

**Comment by MeganerdNL** [comment]
> because of recent changes to the cross toolchains, only this is needed:
> 
> ```diff
>   if [ "$XBPS_TARGET_NO_ATOMIC8" ]; then
>   	XBPS_CROSS_RUSTFLAGS+=" -latomic"
> - 	makedepends+=" libatomic-devel"
>   fi
> - 
> - case "$XBPS_TARGET_MACHINE" in
> - 	*-musl) makedepends+=" musl-devel libstdc++";;
> - esac
> ```
> 
> (tested with a musl masterdir (`xbps-src -m masterdir-musl binary-bootstrap x86_64-musl` and `xbps-src -m masterdir-musl -a armv6l-musl pkg encrypted-dns`))
> 
> > I think that the creation of a mandatory configfile that has an example is quite important
> 
> by important I mean things like breaking changes (see pipewire, for example). things like optional dependencies or getting-started steps are not important enough and just add more noise that users will probably ignore.

Yes, it worked with the new RUSTFLAGS!

I see about the important messages, thanks.

Please review.

**Comment by Duncaen** [comment]
Why is `undelegated.txt` in `/var` instead of `/etc`?

**Comment by MeganerdNL** [comment]
> Why is `undelegated.txt` in `/var` instead of `/etc`?

It's not a config file for the user to edit.

**Comment by Duncaen** [comment]
> > Why is `undelegated.txt` in `/var` instead of `/etc`?
> 
> It's not a config file for the user to edit.

Then I don't understand why its marked mutable, it looks like a configuration file to me.

**Comment by MeganerdNL** [comment]
It's distributed with the package, so mutable (that's how I understand in the manual). Of course it _can_ be edited by the user, if it wants to. Should I put in etc then?

**Comment by Duncaen** [comment]
I would generally stick with what upstream does, which is to use `/etc`.

`mutable_files` basically just disables `xbps-pkgdb` hash checks for the specified, package updates will still overwrite them.

This is generally not useful for most packages and the manual description should be more clear about it.

**Comment by MeganerdNL** [comment]
> I would generally stick with what upstream does, which is to use `/etc`.
> 
> `mutable_files` basically just disables `xbps-pkgdb` hash checks for the specified, package updates will still overwrite them.
> 
> This is generally not useful for most packages and the manual description should be more clear about it. Its more likely to be used for things that are self updating and not to allow the user to make changes as opposed to configuration files.

Thanks for the clearing up, I understand.
undelegated.txt is now in /etc/encrypted-dns in latest push.
